File: PatronesComportamiento/memento.py

class GameMemento:
    def __init__(self, level, health):
        self.level = level
        self.health = health

class Player:
    def __init__(self):
        self.level = 1
        self.health = 100

    def play(self):
        self.level += 1
        self.health -= 10
        print(f" Jugando... Nivel: {self.level}, Vida: {self.health}")

    def save(self):
        print(" Guardando progreso...")
        return GameMemento(self.level, self.health)

    def restore(self, memento):
        self.level = memento.level
        self.health = memento.health
        print(f" Restaurado → Nivel: {self.level}, Vida: {self.health}")


class GameHistory:
    def __init__(self):
        self.saves = []

    def add(self, memento):
        self.saves.append(memento)

    def undo(self):
        if self.saves:
            return self.saves.pop()
        return None



player = Player()
history = GameHistory()


player.play()   
history.add(player.save())   


player.play()   
player.play()   

player.restore(history.undo())

File: PatronesComportamiento/Memento.java
package PatronesComportamiento;


final class GameMemento {
    private final int level;
    private final int health;

    public GameMemento(int level, int health) {
        this.level = level;
        this.health = health;
    }

    public int getLevel() {
        return level;
    }

    public int getHealth() {
        return health;
    }
}


class Player {
    private int level = 1;
    private int health = 100;

    public void play() {
        level++;
        health -= 10;
        System.out.println(" Jugando... Nivel: " + level + " | Vida: " + health);
    }

    public GameMemento save() {
        System.out.println("Guardando progreso...");
        return new GameMemento(level, health);
    }

    public void restore(GameMemento memento) {
        this.level = memento.getLevel();
        this.health = memento.getHealth();
        System.out.println(" Progreso restaurado → Nivel: " + level + " | Vida: " + health);
    }
}


public class Memento {
    public static void main(String[] args) {
        Player player = new Player();

        
        player.play();
        player.play();

        
        GameMemento checkpoint = player.save();

        
        player.play();
        player.play();

        player.restore(checkpoint);
    }
}
